### PR TITLE
Add YOLOv4

### PR DESCRIPTION
**Pull request by jennifererwangg**
# YOLOv4

## Description
[YOLOv4](https://github.com/hunglc007/tensorflow-yolov4-tflite) optimizes the speed and accuracy of object detection. It is twice faster than EfficientDet and improves YOLOv3's AP and FPS by 10% and 12%, respectively. 

## Model
|Model        |Download  |Download (with sample test data)|ONNX version|Opset version|Accuracy |
|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|
|YOLOv4       |[251 MB](model/yolov4.onnx) |[236 MB](model/yolov4.tar.gz)|1.6 |11 |mAP of 0.5733 |

### Source
Tensorflow => onnx model

### Input to model
Input images are resized to the shape `(1, 416, 416, 3)`.

### Preprocessing steps
The following code shows how to preprocess an image. For more information and an example on how preprocess is done, please visit [this](dependencies/ort.ipynb) notebook.

```python
import numpy as np
import cv2

# this function is from tensorflow-yolov4-tflite/core/utils.py
def image_preprocess(image, target_size, gt_boxes=None):

    ih, iw    = target_size
    h,  w, _  = image.shape

    scale = min(iw/w, ih/h)
    nw, nh  = int(scale * w), int(scale * h)
    image_resized = cv2.resize(image, (nw, nh))

    image_paded = np.full(shape=[ih, iw, 3], fill_value=128.0)
    dw, dh = (iw - nw) // 2, (ih-nh) // 2
    image_paded[dh:nh+dh, dw:nw+dw, :] = image_resized
    image_paded = image_paded / 255.

    if gt_boxes is None:
        return image_paded

    else:
        gt_boxes[:, [0, 2]] = gt_boxes[:, [0, 2]] * scale + dw
        gt_boxes[:, [1, 3]] = gt_boxes[:, [1, 3]] * scale + dh
        return image_paded, gt_boxes

# input
input_size = 416

original_image = cv2.imread("input.jpg")
original_image = cv2.cvtColor(original_image, cv2.COLOR_BGR2RGB)
original_image_size = original_image.shape[:2]

image_data = image_preprocess(np.copy(original_image), [input_size, input_size])
image_data = image_data[np.newaxis, ...].astype(np.float32)

```

### Output of model
Output shape: `(1, 52, 52, 3, 85)`

There are 3 output layers. For each layer, there are 255 outputs: 85 values per anchor, times 3 anchors.

The 85 values of each anchor consists of 4 box coordinates describing the predicted bounding box (x, y, h, w), 1 object confidence, and 80 class confidences. [Here](https://github.com/hunglc007/tensorflow-yolov4-tflite/blob/master/data/classes/coco.names) is the class list.

### Dataset (Train and validation)
Pretrained yolov4 weights can be downloaded [here](https://drive.google.com/open?id=1cewMfusmPjYWbrnuJRuKhPMwRe_b9PaT).

### Validation accuracy
mAP50 on COCO 2017 dataset is 0.5733, based on the original tensorflow [model](https://github.com/hunglc007/tensorflow-yolov4-tflite#map50-on-coco-2017-dataset).
<hr>

## References
This model is diretly coverted from [hunglc007/tensorflow-yolov4-tflite](https://github.com/hunglc007/tensorflow-yolov4-tflite).

## License
MIT License

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -68,6 +68,7 @@ Object detection models detect the presence of multiple objects in an image and
 |<b>[YOLO v2-coco](vision/object_detection_segmentation/yolov2-coco)</b>|[Redmon et al.](https://arxiv.org/abs/1612.08242)|A CNN model for real-time object detection system that can detect over 9000 object categories. It uses a single network evaluation, enabling it to be more than 1000x faster than R-CNN and 100x faster than Faster R-CNN. This model is trained with COCO dataset and contains 80 classes.
 |<b>[YOLO v3](vision/object_detection_segmentation/yolov3)</b>|[Redmon et al.](https://arxiv.org/pdf/1804.02767.pdf)|A deep CNN model for real-time object detection that detects 80 different classes. A little bigger than YOLOv2 but still very fast. As accurate as SSD but 3 times faster.|
 |<b>[Tiny YOLOv3](vision/object_detection_segmentation/tiny-yolov3)</b>|[Redmon et al.](https://arxiv.org/pdf/1804.02767.pdf)| A smaller version of YOLOv3 model. |
+|<b>[YOLOv4](vision/object_detection_segmentation/yolov4)</b>|[Bochkovskiy et al.](https://arxiv.org/abs/2004.10934)|Optimizes the speed and accuracy of object detection. Twice faster than EfficientDet and improves YOLOv3's AP and FPS by 10% and 12%, respectively. |
```
Comment (vinitra) [inline README.md:71]: wording suggestions:
Two times faster than EfficientDet
define AP and FPS (see mIOU in DUC description)

File: vision/object_detection_segmentation/yolov4/README.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+## Description
+[YOLOv4](https://github.com/hunglc007/tensorflow-yolov4-tflite) optimizes the speed and accuracy of object detection. It is twice faster than EfficientDet and improves YOLOv3's AP and FPS by 10% and 12%, respectively. 
+
+## Model
+
+|Model        |Download  |Download (with sample test data)|ONNX version|Opset version|Accuracy |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|YOLOv4       |[251 MB](model/yolov4.onnx) |[236 MB](model/yolov4.tar.gz)|1.6 |11 |mAP of 0.5733 |
+
+
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/README.md:12]: Add section for where this model is converted from
Tensorflow -> ONNX

File: vision/object_detection_segmentation/yolov4/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+## Inference
+### Conversion
+A tutorial for the conversion process can be found in the [conversion](dependencies/conversion.ipynb) notebook.
+
+Validation of the converted model and a graph representation of it can be found in the [validation](dependencies/onnx-model-validation.ipynb) notebook.
+### Running inference
+A tutorial for running inference using onnxruntime can be found in the [ort](dependencies/ort.ipynb) notebook.
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/README.md:20]: rename to inference notebook

File: vision/object_detection_segmentation/yolov4/README.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+### Input to model
+Input images are resized to the shape `(1, 416, 416, 3)`.
+
+### Preprocessing steps
+The following code shows how to preprocess an image. For more information and an example on how preprocess is done, please visit [this](dependencies/ort.ipynb) notebook.
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/README.md:25]: nit: how preprocessing is done
nit: please visit the [inference] notebook

File: vision/object_detection_segmentation/yolov4/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+## Inference
+### Conversion
+A tutorial for the conversion process can be found in the [conversion](dependencies/conversion.ipynb) notebook.
+
+Validation of the converted model and a graph representation of it can be found in the [validation](dependencies/onnx-model-validation.ipynb) notebook.
+### Running inference
+A tutorial for running inference using onnxruntime can be found in the [ort](dependencies/ort.ipynb) notebook.
+### Input to model
+Input images are resized to the shape `(1, 416, 416, 3)`.
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/README.md:22]: Describe the input dimensions. Likely something like -> (Batch-size, height, width, channels) but confirm.

File: vision/object_detection_segmentation/yolov4/README.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+### Output of model
+Output shape: `(1, 52, 52, 3, 85)`
+
+There are 3 output layers. For each layer, there are 255 outputs: 85 values per anchor, times 3 anchors.
+
+The 85 values of each anchor consists of 4 box coordinates describing the predicted bounding box (x, y, h, w), 1 object confidence, and 80 class confidences. [Here](https://github.com/hunglc007/tensorflow-yolov4-tflite/blob/master/data/classes/coco.names) is the class list.
+
+### Postprocessing steps
+To see postprocess steps and an example, please visit [this](dependencies/ort.ipynb) notebook.
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/README.md:74]: nit: postprocessing steps
nit: the [inference] notebook

File: vision/object_detection_segmentation/yolov4/README.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+[YOLOv4](https://github.com/hunglc007/tensorflow-yolov4-tflite) optimizes the speed and accuracy of object detection. It is two times faster than EfficientDet. It improves YOLOv3's AP and FPS by 10% and 12%, respectively, with mAP50 of 52.32 on the COCO 2017 dataset and FPS of 41.7 on Tesla 100.
+
+## Model
+
+|Model        |Download  |Download (with sample test data)|ONNX version|Opset version|Accuracy |
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|YOLOv4       |[251 MB](model/yolov4.onnx) |[236 MB](model/yolov4.tar.gz)|1.6 |11 |mAP of 0.5733 |
+
+### Source
+Tensorflow => ONNX
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/README.md:13]:
```suggestion
### Source
Tensorflow YOLOv4 => ONNX YOLOv4
```

File: vision/object_detection_segmentation/yolov4/README.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+## Dataset
+Pretrained yolov4 weights can be downloaded [here](https://drive.google.com/open?id=1cewMfusmPjYWbrnuJRuKhPMwRe_b9PaT). 
+
+## Validation accuracy
+mAP50 on COCO 2017 dataset is 0.5733, based on the original tensorflow [model](https://github.com/hunglc007/tensorflow-yolov4-tflite#map50-on-coco-2017-dataset).
+
+## Publication/Attribution
+* [YOLOv4: Optimal Speed and Accuracy of Object Detection](https://arxiv.org/abs/2004.10934). Alexey Bochkovskiy, Chien-Yao Wang, Hong-Yuan Mark Liao. 
+* Original [models](https://github.com/AlexeyAB/darknet).
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/README.md:86]:
```suggestion
* Original models from [Darknet Github repository](https://github.com/AlexeyAB/darknet).
```

File: vision/object_detection_segmentation/yolov4/dependencies/inference.ipynb [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+    "from onnx import numpy_helper\n",
+    "import onnx\n",
+    "import os\n",
+    "\n",
+    "data_dir = \"./test_data_set\"\n",
+    "test_data_dir = os.path.join(data_dir)\n",
+    "if not os.path.exists(test_data_dir):\n",
+    "    os.makedirs(test_data_dir)\n",
+    "        \n",
+    "# Convert the Numpy array to a TensorProto\n",
```
Comment (vinitra) [inline vision/object_detection_segmentation/yolov4/dependencies/inference.ipynb:133]: nit: NumPy
